### ISSUE TITLE
Charmrun should test for SSH failures when node programs fail to launch

### ISSUE THREAD

**Issue by PhilMiller** [UIUC-PPL/charm#789]
*Original issue: https://charm.cs.illinois.edu/redmine/issues/789*

---
When charmrun sees a failure in ssh-ing to a compute node, the fault could either be in the SSH connection or in the node script or compute process itself. To distinguish these cases, charmrun could try a really simple ssh connection running something like `echo hello`. If that works, SSH is happy and the problem is in the script or executable; if it fails, then we can unambiguously report that the problem lies with SSH.

**Comment by PhilMiller** [comment]
*Original date: 2017-04-11 20:58:04*

---
It'd be nice to have, but nothing is broken by missing this.

**Comment by evan-charmworks** [comment]
*Original date: 2018-01-17 20:51:41*

---
Would doing this imply that we create an SSH connection just to run `echo hello` or similar as a diagnostic, then disconnect, create a new one, and proceed as before? If so, this has the potential to double the time it takes to launch. If not, I don't immediately understand how we could achieve the separation of fault domains.

**Comment by stwhite91** [comment]
*Original date: 2018-01-17 21:04:09*

---
I think the idea is to flip the ordering of what you said: first try to actually connect, then if there is an error do the 'echo hello' test to narrow down the cause of the error

**Comment by evan-charmworks** [comment]
*Original date: 2018-01-23 22:17:47*

---
The following lines are printed by the script run on the remote ssh instance with ++verbose:

```
Charmrun remote shell(localhost1.0)> remote responding...
Charmrun remote shell(localhost1.0)> starting node-program...
Charmrun remote shell(localhost1.0)> remote shell phase successful.
```

I believe the presence or absence of these lines should be sufficient to diagnose the problem as described in this issue. Feedback?

**Comment by evan-charmworks** [comment]
*Original date: 2018-02-07 20:54:29*

---
Removing version tag because I consider this issue satisfied, but am waiting on feedback before marking it closed.

**Comment by evan-charmworks** [comment]
*Original date: 2018-05-10 22:40:26*

---
++verbose is definitely enough to diagnose the described problem, but I've added this fact to the documentation in order to satisfy this issue: ~~https://charm.cs.illinois.edu/gerrit/4163~~ https://github.com/UIUC-PPL/charm/commit/e264b8813a46bdf48d366100ff47bcbf80911e3b
